pipeline/trusted-resources/pkg/trustedtask/trustedpipelinerun.go

```diff
@@ -0,0 +1,188 @@
+				continue
+			}
+
+			fn, err := taskrun.GetTaskFunc(ctx, k8sclient, tektonClient, t.TaskRef, pr.Namespace, serviceAccountName)
+			if err != nil {
+				return apis.ErrGeneric(err.Error(), "spec", "taskRef")
+			}
+
+			if err := verifyTask(ctx, t.TaskRef.Name, k8sclient, fn); err != nil {
+				return apis.ErrGeneric(err.Error(), "spec", "taskRef")
+			}
+		}
+		return nil
+	}
+
+	cfg := config.FromContextOrDefaults(ctx)
+	cfg.FeatureFlags.EnableTektonOCIBundles = true
+	ctx = config.ToContext(ctx, cfg)
+
+	fn, err := resources.GetPipelineFunc(ctx, k8sclient, tektonClient, &pr.PipelineRun)
+	if err != nil {
+		return apis.ErrGeneric(err.Error(), "spec", "PipelineRef")
+	}
+	resolvedPipeline, err := fn(ctx, pr.Spec.PipelineRef.Name)
+	if err != nil {
+		return apis.ErrGeneric(err.Error(), "spec", "PipelineRef")
+	}
+
+	pipeline, signature, err := preparePipeline(resolvedPipeline)
+	if err != nil {
+		return apis.ErrGeneric(err.Error(), "spec", "PipelineRef")
+	}
+
+	v, err := verifier(ctx, pipeline.ObjectMeta.Annotations, k8sclient)
+	if err != nil {
+		return apis.ErrGeneric(err.Error(), "PipelineRef")
+	}
+
+	if err := VerifyInterface(ctx, pipeline, v, signature); err != nil {
+		return apis.ErrGeneric(err.Error(), "PipelineRef")
+	}
+
+	for _, t := range pipeline.Spec.Tasks {
+		if t.TaskRef == nil {
+			continue
+		}
+
+		fn, err := taskrun.GetTaskFunc(ctx, k8sclient, tektonClient, t.TaskRef, pr.Namespace, serviceAccountName)
+		if err != nil {
+			return apis.ErrGeneric(err.Error(), "spec", "taskRef")
+		}
+
+		if err := verifyTask(ctx, t.TaskRef.Name, k8sclient, fn); err != nil {
+			return apis.ErrGeneric(err.Error(), "spec", "taskRef")
+		}
+
+	}
+
+	return nil
+}
+
+// preparePipeline will convert the taskobject to task and extract the signature.
+func preparePipeline(p v1beta1.PipelineObject) (v1beta1.Pipeline, []byte, error) {
+	pipeline := v1beta1.Pipeline{}
+	pipeline.TypeMeta = metav1.TypeMeta{
+		APIVersion: "tekton.dev/v1beta1",
+		Kind:       "Pipeline"}
+	pipeline.Spec = p.PipelineSpec()
+
+	// exclude the fields populated by system.
+	pipeline.Name = p.PipelineMetadata().Name
+	pipeline.GenerateName = p.PipelineMetadata().GenerateName
+	pipeline.Namespace = p.PipelineMetadata().Namespace
+
+	pipeline.Labels = make(map[string]string)
+	for k, v := range p.PipelineMetadata().Labels {
+		pipeline.Labels[k] = v
+	}
+
+	pipeline.Annotations = make(map[string]string)
+	for k, v := range p.PipelineMetadata().Annotations {
+		pipeline.Annotations[k] = v
+	}
+
+	// exclude the annotations added by other components
+	delete(pipeline.ObjectMeta.Annotations, "kubectl-client-side-apply")
+	delete(pipeline.ObjectMeta.Annotations, "kubectl.kubernetes.io/last-applied-configuration")
+
+	// extract signature
+	signature, err := base64.StdEncoding.DecodeString(pipeline.ObjectMeta.Annotations[SignatureAnnotation])
+	if err != nil {
+		return pipeline, nil, err
+	}
+	delete(pipeline.ObjectMeta.Annotations, SignatureAnnotation)
```

Comment:
It would be great if we could consolidate this with the taskrun code, since it's effectively the same operations on the objectmeta. Any ideas here? 🤔

Reply:
just merged in the latest commit